### PR TITLE
feat: expose well known env vars as part of core

### PR DESCRIPTION
**Pull request by peternhale**
@W-9838863@

### CODE REVIEW

File: messages/envVars.md [hunk 1]
```diff
@@ -0,0 +1,255 @@
+# forceOpenUrl
```
Comment (peternhale) [inline messages/envVars.md:1]: message file for all env vars. Needs lots of work.

File: src/config/configAggregator.ts [hunk 2]
```diff
@@ -344,10 +343,10 @@ export class ConfigAggregator extends AsyncOptionalCreatable<JsonMap> {
   }
 
   private resolveProperties(globalConfig: JsonMap, localConfig?: JsonMap): JsonMap {
-    const accumulator: Dictionary<string> = {};
+    const accumulator = envVarsResolve();
```
Comment (peternhale) [inline src/config/configAggregator.ts:346]: Start wil resolved env vars.

File: src/config/configAggregator.ts [hunk 3]
```diff
@@ -344,10 +343,10 @@ export class ConfigAggregator extends AsyncOptionalCreatable<JsonMap> {
...
     this.setEnvVars(
       this.getAllowedProperties().reduce((obj, property) => {
-        const val = process.env[propertyToEnvName(property.key)];
+        const val = accumulator[propertyToEnvName(property.key)];
```
Comment (peternhale) [inline src/config/configAggregator.ts:349]: Look to resolved env vars rather than process.env

File: src/exported.ts [hunk 4]
```diff
@@ -51,7 +62,7 @@ export { Messages } from './messages';
 
 export { Org } from './org/org';
 
-export { OrgConfigProperties } from './org/orgConfigProperties';
+export { OrgConfigProperties, ORG_CONFIG_ALLOWED_PROPERTIES } from './org/orgConfigProperties';
```
Comment (peternhale) [inline src/exported.ts:65]: Gives devs access to this constant to access description of config var

File: src/globalInfo/index.ts [hunk 5]
```diff
@@ -8,6 +8,7 @@
 export * from './globalInfoConfig';
 export * from './sfdxDataHandler';
 export * from './types';
+export { DxWellKnownEnvVars, DX_SUPPORTED_ENV_VARS } from './dxEnvVars';
```
Comment (peternhale) [inline src/globalInfo/index.ts:11]: Give devs access to constants

File: src/org/orgConfigProperties.ts [hunk 6]
```diff
@@ -13,10 +19,10 @@ export enum OrgConfigProperties {
 export const ORG_CONFIG_ALLOWED_PROPERTIES = [
   {
     key: OrgConfigProperties.TARGET_ORG,
-    description: 'The target to be used for any command communicating with an org.',
+    description: messages.getMessage('targetOrg'),
```
Comment (peternhale) [inline src/org/orgConfigProperties.ts:22]: Use message file based descriptions

File: src/globalInfo/dxEnvVars.ts [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+  'SF_USE_PROGRESS_BAR' = 'SF_USE_PROGRESS_BAR',
+}
+type EnvMetaData = {
+  description: string;
+  synonymOf: Optional<string>;
+};
+type EnvType = {
+  [key: string]: EnvMetaData;
+};
+export const DX_SUPPORTED_ENV_VARS: EnvType = {
```
Comment (peternhale) [inline src/globalInfo/dxEnvVars.ts:88]: map of env var metadata

File: src/globalInfo/dxEnvVars.ts [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+  },
+  [DxWellKnownEnvVars.SF_USE_PROGRESS_BAR]: {
+    description: messages.getMessage(camelCase(DxWellKnownEnvVars.SF_USE_PROGRESS_BAR)),
+    synonymOf: undefined,
+  },
+};
+
+export const envVarsResolve = (): Dictionary<string> => {
+  const dict = {} as Dictionary<string>;
+  Object.entries(process.env).forEach(([key, value]) => {
```
Comment (peternhale) [inline src/globalInfo/dxEnvVars.ts:355]: Assume control of pulling in process.env from config agg

File: src/globalInfo/dxEnvVars.ts [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+    // save all env vars to dictionary
+    dict[key] = value;
```
Comment (peternhale) [inline src/globalInfo/dxEnvVars.ts:357]: All env vars are saved in the dictionary

File: src/globalInfo/dxEnvVars.ts [hunk 10]
```diff
@@ -0,0 +1,371 @@
...
+    // cross populate value to synonym if synonym is null or undefined
+    if (DX_SUPPORTED_ENV_VARS[key]) {
+      if (DX_SUPPORTED_ENV_VARS[key].synonymOf) {
+        const synonym = DX_SUPPORTED_ENV_VARS[key].synonymOf;
+        // set synonym only if it is not set in in current env
+        if (synonym && Global.SFDX_INTEROPERABILITY) {
```
Comment (peternhale) [inline src/globalInfo/dxEnvVars.ts:363]: Only apply to synonym if interop is enabled

File: src/globalInfo/dxEnvVars.ts [hunk 11]
```diff
@@ -0,0 +1,371 @@
...
+export const envVarsResolve = (): Dictionary<string> => {
+  const dict = {} as Dictionary<string>;
+  Object.entries(process.env).forEach(([key, value]) => {
+    // save all env vars to dictionary
+    dict[key] = value;
+    // cross populate value to synonym if synonym is null or undefined
+    if (DX_SUPPORTED_ENV_VARS[key]) {
+      if (DX_SUPPORTED_ENV_VARS[key].synonymOf) {
+        const synonym = DX_SUPPORTED_ENV_VARS[key].synonymOf;
+        // set synonym only if it is not set in in current env
```
Comment (peternhale) [inline src/globalInfo/dxEnvVars.ts:362]: Need to fix this comment

File: src/globalInfo/dxEnvVars.ts [hunk 12]
```diff
@@ -0,0 +1,371 @@
...
+    synonymOf: undefined,
+  },
+};
+
+export const envVarsResolve = (): Dictionary<string> => {
+  const dict = {} as Dictionary<string>;
+  Object.entries(process.env).forEach(([key, value]) => {
+    // save all env vars to dictionary
+    dict[key] = value;
+    // cross populate value to synonym if synonym is null or undefined
```
Comment (peternhale) [inline src/globalInfo/dxEnvVars.ts:358]: need to fix this comment

File: src/globalInfo/dxEnvVars.ts [hunk 13]
```diff
@@ -0,0 +1,371 @@
...
+    if (DX_SUPPORTED_ENV_VARS[key]) {
+      if (DX_SUPPORTED_ENV_VARS[key].synonymOf) {
+        const synonym = DX_SUPPORTED_ENV_VARS[key].synonymOf;
+        // set synonym only if it is not set in in current env
+        if (synonym && Global.SFDX_INTEROPERABILITY) {
+          dict[synonym] = process.env[key];
+          process.env[synonym] = dict[synonym];
```
Comment (peternhale) [inline src/globalInfo/dxEnvVars.ts:365]: In addition to placing the synonym value into the dictionary, also set process.env

File: src/globalInfo/dxEnvVars.ts [hunk 14]
```diff
@@ -0,0 +1,371 @@
...
+  },
+  [DxWellKnownEnvVars.SF_CODE_COVERAGE_REQUIREMENT]: {
+    description: messages.getMessage(camelCase(DxWellKnownEnvVars.SF_CODE_COVERAGE_REQUIREMENT)),
+    synonymOf: undefined,
+  },
+  [DxWellKnownEnvVars.SF_CONTENT_TYPE]: {
+    description: messages.getMessage(camelCase(DxWellKnownEnvVars.SF_CONTENT_TYPE)),
+    synonymOf: undefined,
+  },
+  [DxWellKnownEnvVars.SF_DEFAULTDEVHUBUSERNAME]: {
```
Comment (mdonnalley) [inline src/globalInfo/dxEnvVars.ts:261]: `SF_TARGET_DEV_HUB`

File: src/globalInfo/dxEnvVars.ts [hunk 15]
```diff
@@ -0,0 +1,371 @@
...
+    description: messages.getMessage(camelCase(DxWellKnownEnvVars.SF_DEFAULTDEVHUBUSERNAME)),
+    synonymOf: undefined,
+  },
+  [DxWellKnownEnvVars.SF_DEFAULTUSERNAME]: {
```
Comment (mdonnalley) [inline src/globalInfo/dxEnvVars.ts:265]: `SF_TARGET_ORG`

File: test/unit/globalInfo/dxEnvVars.test.ts [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+  it('should load well known env var and set synonym', () => {
+    process.env['SFDX_ACCESS_TOKEN'] = 'some access token';
+    const envVars = envVarsResolve();
+    expect(envVars['SFDX_ACCESS_TOKEN']).to.equal('some access token');
+    expect(envVars['SF_ACCESS_TOKEN']).to.equal('some access token');
+  });
+  it('should load well known env var not set synonym interoperability is disabled', () => {
+    process.env['SFDX_ACCESS_TOKEN'] = 'some access token';
+    process.env['SF_ACCESS_TOKEN'] = 'some other access token';
+    process.env['SF_SSFDX_INTEROPERABILITY'] = 'false';
```
Comment (mdonnalley) [inline test/unit/globalInfo/dxEnvVars.test.ts:34]: `SF_SSFDX_INTEROPERABILITY` => `SF_SFDX_INTEROPERABILITY`

File: src/config/configAggregator.ts (hunk 2 above)
Comment (mdonnalley) [reply src/config/configAggregator.ts:346]: Does it make sense to expose a more elaborate abstraction? Something like this:

```typescript
const envVars = new EnvVars();

envVars.get(propertyToEnvName(property.key))
```

And then we could have other useful methods, e.g. `all`, `set`, `unset`, etc...

File: src/globalInfo/dxEnvVars.ts [hunk 17]
```diff
@@ -0,0 +1,371 @@
...
+    description: messages.getMessage(camelCase(DxWellKnownEnvVars.SFDX_IMPROVED_CODE_COVERAGE)),
+    synonymOf: DxWellKnownEnvVars.SF_IMPROVED_CODE_COVERAGE,
+  },
+  [DxWellKnownEnvVars.SFDX_INSTANCE_URL]: {
+    description: messages.getMessage(camelCase(DxWellKnownEnvVars.SFDX_INSTANCE_URL)),
+    synonymOf: DxWellKnownEnvVars.SF_INSTANCE_URL,
+  },
+  [DxWellKnownEnvVars.SFDX_JSON_TO_STDOUT]: {
+    description: messages.getMessage(camelCase(DxWellKnownEnvVars.SFDX_JSON_TO_STDOUT)),
+    synonymOf: DxWellKnownEnvVars.SF_JSON_TO_STDOUT,
```
Comment (mdonnalley) [inline src/globalInfo/dxEnvVars.ts:180]: We should go through these and think carefully about whether it needs to map to `sf` or not. This one for example would have no effect on `sf`

Comment (peternhale) [reply src/globalInfo/dxEnvVars.ts:180]: @mdonnalley The list of `SF` env vars is speculative at this moment, since I son't find any references to `SF` env vars in core. I will pull these out into a doc for further discussion.

File: src/config/configAggregator.ts [hunk 18]
```diff
@@ -78,14 +78,14 @@ export interface ConfigInfo {
  * ```
  */
 export class ConfigAggregator extends AsyncOptionalCreatable<JsonMap> {
-  private static instance: AsyncOptionalCreatable;
+  private static instance: AsyncOptionalCreatable | undefined;
```
Comment (mdonnalley) [inline src/config/configAggregator.ts:81]: Why did this need to change?

**Review by mdonnalley — APPROVED** [review]

File: src/config/configAggregator.ts (hunk 18 above)
Comment (peternhale) [reply src/config/configAggregator.ts:81]: @mdonnalley I will look, but I recall hitting a test failure where a call is being made directly to buildJwt... function, without calling create or new on a ConfigAggregator.

**Comment by peternhale** [comment]
@RodEsp Looking for your feedback on this PR, when you have a moment.

File: src/config/configAggregator.ts [hunk 19]
```diff
@@ -321,11 +324,7 @@ export class ConfigAggregator extends AsyncOptionalCreatable<JsonMap> {
    * Add an allowed config property.
    */
   public addAllowedProperties(configMetas: ConfigPropertyMeta | ConfigPropertyMeta[]): void {
-    if (isArray(configMetas)) {
-      this.allowedProperties.push(...configMetas);
-    } else {
-      this.allowedProperties.push(configMetas);
-    }
+    this.allowedProperties.push(...(isArray(configMetas) ? configMetas : [configMetas]));
```
Comment (RodEsp) [inline src/config/configAggregator.ts:327]: Why this change? It doesn't change the functionality but makes the code harder to follow.

Is there a performance benefit or something like that?

File: src/globalInfo/dxEnvVars.ts [hunk 20]
```diff
@@ -0,0 +1,394 @@
...
+  synonymOf: Nullable<string>;
+};
+type EnvType = {
+  [key in EnvironmentVariable]: EnvMetaData;
+};
+
+const getMessage = (environmentVariable: EnvironmentVariable): string =>
+  messages.getMessage(camelCase(environmentVariable));
+
+export const DX_SUPPORTED_ENV_VARS: EnvType = {
```
Comment (RodEsp) [inline src/globalInfo/dxEnvVars.ts:90]: Why the `DX_` prefix?
Can we call this `SUPPORTED_ENV_VARS` and this file just `envVars.ts` instead of having `DX_` and `dx` as a prefix for the sake of simplicity or do you think it adds anything worthwhile?

File: src/globalInfo/dxEnvVars.ts [hunk 21]
```diff
@@ -0,0 +1,394 @@
...
+    description: getMessage(EnvironmentVariable.SF_USE_GENERIC_UNIX_KEYCHAIN),
+    synonymOf: null,
+  },
+  [EnvironmentVariable.SF_USE_PROGRESS_BAR]: {
+    description: getMessage(EnvironmentVariable.SF_USE_PROGRESS_BAR),
+    synonymOf: null,
+  },
+};
+
+export class EnvVars extends Env {
```
Comment (RodEsp) [inline src/globalInfo/dxEnvVars.ts:343]: This class and methods should probably live in an `envVarAccessor` to maintain the pattern we're using with [aliases](https://github.com/forcedotcom/sfdx-core/blob/v3/src/globalInfo/accessors/aliasAccessor.ts), [orgs](https://github.com/forcedotcom/sfdx-core/blob/v3/src/globalInfo/accessors/orgAccessor.ts), and [tokens](https://github.com/forcedotcom/sfdx-core/blob/v3/src/globalInfo/accessors/tokenAccessor.ts).

**Review by RodEsp — CHANGES_REQUESTED** [review]

File: src/config/configAggregator.ts (hunk 19 above)
Comment (peternhale) [reply src/config/configAggregator.ts]: reverted

File: src/globalInfo/dxEnvVars.ts (hunk 20 above)
Comment (peternhale) [reply src/globalInfo/dxEnvVars.ts]: Removed dx prefix

File: src/globalInfo/dxEnvVars.ts (hunk 21 above)
Comment (peternhale) [reply src/globalInfo/dxEnvVars.ts]: I was wondering where the best place to keep this class. I don't feel that it qualifies as an accessor and since it is closely tied to config aggregator, I moved the class to config.

File: src/config/envVars.ts [hunk 22]
```diff
@@ -0,0 +1,389 @@
...
+
+  public asMap(): Map<string, string> {
+    return this.entries().reduce((accumulator, [key, value]) => {
+      accumulator.set(key, value);
+      return accumulator;
+    }, new Map<string, string>());
+  }
+
+  private resolve(): void {
+    const overrideInteroperabilityEnvVar = process.env['TEST_OVERRIDE_GLOBAL_SF_SFDX_INTEROPERABILITY'];
```
Comment (RodEsp) [inline src/config/envVars.ts:371]: Is this needed?

**Review by RodEsp — APPROVED** [review]

File: messages/envVars.md [hunk 23]
```diff
@@ -0,0 +1,239 @@
...
+
+Sets the URL to a private npm server, where all packages that you publish are private.
+
+# sfPrecompileEnable
+
+Set to true to enable Apex pre-compile before the tests are run. Default is false.
+
+# sfProjectAutoupdateDisableForPackageCreate
+
+some useful text
```
Comment (RodEsp) [inline messages/envVars.md:223]: Missing an appropriate message.

File: messages/envVars.md [hunk 24]
```diff
@@ -0,0 +1,239 @@
...
+
+Set to true to disable the CLI from collecting usage information, user environment information, and crash reports.
+
+# sfDnsTimeout
+
+Specifies the number of seconds that the force:org:\* commands wait for a response when checking whether an org is connected. If the commands don’t receive a response in that time, they time out. Default value is 3.
+
+# sfDomainRetry
+
+some useful text
```
Comment (RodEsp) [inline messages/envVars.md:187]: Missing an appropriate message.

**Review by RodEsp — CHANGES_REQUESTED** [review]

**Review by RodEsp — APPROVED** [review]
